eval/topdown_test.go

```diff
@@ -0,0 +1,540 @@
+		}
+
+		switch e := tc.expected.(type) {
+		case string:
+			expected := loadExpectedResult(e)
+			result, err := TopDownQuery(ctx)
+			if err != nil {
+				t.Errorf("Test case %d (%v): unexpected error: %v", i+1, tc.note, err)
+				continue
+			}
+			if !reflect.DeepEqual(result, expected) {
+				t.Errorf("Test case %d (%v): expected %v but got %v", i+1, tc.note, expected, result)
+			}
+		case error:
+			_, err := TopDownQuery(ctx)
+			if !reflect.DeepEqual(err, e) {
+				t.Errorf("Test case %d (%v): expected error %v but got %v", i+1, tc.note, e, err)
+			}
+		}
+
+	}
+}
+
+func TestTopDownEqExpr(t *testing.T) {
+
+	tests := []struct {
+		note     string
+		rule     string
+		expected interface{}
+	}{
+		// undefined cases
+		{"undefined: same type", `p = true :- true = false`, ""},
+		{"undefined: diff type", `p = true :- 42 = "hello"`, ""},
+		{"undefined: array order", `p = true :- [1,2,3] = [1,3,2]`, ""},
+		{"undefined: ref value", "p = true :- a[3] = 9999", ""},
+		{"undefined: ref values", "p = true :- a[i] = 9999", ""},
+		{"undefined: ground var", "p = true :- a[3] = x, x = 3", ""},
+		{"undefined: array var 1", "p = true :- [1,x,x] = [1,2,3]", ""},
+		{"undefined: array var 2", "p = true :- [1,x,3] = [1,2,x]", ""},
+		{"undefined: object var 1", `p = true :- {"a": 1, "b": 2} = {"a": a, "b": a}`, ""},
+		{"undefined: array deep var 1", "p = true :- [[1,x],[3,x]] = [[1,2],[3,4]]", ""},
+		{"undefined: array deep var 2", "p = true :- [[1,x],[3,4]] = [[1,2],[x,4]]", ""},
+		{"undefined: array uneven", `p = true :- [true, false, "foo", "deadbeef"] = c[i][j]`, ""},
+		{"undefined: object uneven", `p = true :- {"a": 1, "b": 2} = {"a": 1}`, ""},
+		{"undefined: occurs 1", "p = true :- [y,x] = [[x],y]", ""},
+		{"undefined: occurs 2", "p = true :- [y,x] = [{\"a\": x}, y]", ""},
+
+		// ground terms
+		{"ground: bool", `p = true :- true = true`, "true"},
+		{"ground: string", `p = true :- "string" = "string"`, "true"},
+		{"ground: number", `p = true :- 17 = 17`, "true"},
+		{"ground: null", `p = true :- null = null`, "true"},
+		{"ground: array", `p = true :- [1,2,3] = [1,2,3]`, "true"},
+		{"ground: object", `p = true :- {"b": false, "a": [1,2,3]} = {"a": [1,2,3], "b": false}`, "true"},
+		{"ground: ref 1", `p = true :- a[2] = 3`, "true"},
+		{"ground: ref 2", `p = true :- b["v2"] = "goodbye"`, "true"},
+		{"ground: ref 3", `p = true :- d["e"] = ["bar", "baz"]`, "true"},
+		{"ground: ref 4", `p = true :- c[0].x[1] = c[0].z["q"]`, "true"},
+
+		// variables
+		{"var: a=b=c", "p[a] :- a = b, c = 42, b = c", "[42]"},
+		{"var: ref value", "p = true :- a[3] = x, x = 4", "true"},
+		{"var: ref values", "p = true :- a[i] = x, x = 2", "true"},
+		{"var: ref key", "p = true :- a[i] = 4, x = 3", "true"},
+		{"var: ref keys", "p = true :- a[i] = x, i = 2", "true"},
+		{"var: ref ground var", "p[x] :- i = 2, a[i] = x", "[3]"},
+		{"var: ref ref", "p[x] :- c[0].x[i] = c[0].z[j], x = [i, j]", `[[0, "p"], [1, "q"]]`},
+
```

Comment:
Is it possible to add some tests for multiple rules with the same head?

p[x] :- a[x]
p[x] :- b[x]